DOCS-1472:  Add navigation and teleoperation use case (#2341)
viamrobotics/docs: docs/use-cases/navigate.md

---

> title: "Teleoperate and navigate with rovers"

Reviewer: I think teleoperate is a transitive verb and navigate is used with the transitive meaning - you're navigating the rover (with your Viam-powered software), so I think this was correct as is. What is your thinking?

Author: Oh that does sound correct, sorry! Fixing

> title: "Teleoperate and navigate with rovers"

Author: ```suggestion
title: "Teleoperate and navigate rovers"
```

linkTitle: "Teleoperate and navigate with rovers"

> weight: 50

Reviewer: why are you changing this?

> weight: 50

Author: ```suggestion
weight: 60
```

type: "docs"
tags: ["navigation", "motion", "services", "base", "base remote control"]
no_list: true
description: "Use Viam to teleoperate and navigate rover bases."
image: "/services/icons/navigation.svg"
imageAlt: "Navigation"
images: ["/services/icons/navigation.svg"]
---

> If you have a rover base, you can teleoperate and navigate it with Viam.

Reviewer: Either works

> If you have a rover base, you can teleoperate and navigate it with Viam.

Author: Your suggestion sounds good!

> Remotely control your base after configuring your machine on the app's **Control** tab, and set up autonomous navigation with the [navigation service](/mobility/navigation/).

Reviewer 2: ```suggestion
```
I *think* this sentence could actually be removed since it's sort of just a reiteration and expansion of the first sentence, but you're going to cover all these points below too, so don't need to get into the details here.

You might consider adding a sentence about why you'd want to do this with Viam, such as that all these tools are integrated seamlessly, or something.

> Remotely control your base after configuring your machine on the app's **Control** tab, and set up autonomous navigation with the [navigation service](/mobility/navigation/).

Reviewer: Currently sounds like you're configuring the machine on the control tab:
```suggestion
Once you have configured your machine, you can remotely control your machine on the app's **Control** tab, and set up autonomous navigation with the [navigation service](/mobility/navigation/).
```

<table>
  <tr>
    <th>{{<imgproc src="/use-cases/base-control.png" class="fill alignright" resize="200x" style="max-width: 200px" declaredimensions=true alt="Base control card">}}

>       <b>1. Teleoperate</b>

Reviewer 2: ```suggestion
      <b>1. Teleoperate</b><br><br>
```
[opt]

Reviewer: If we change this, please make a follow up ticket to ensure all  use case pages are consistent! Thanks!

Author: Think it looks ok as is

Reviewer 2: Okay--right now two of the live use cases use br br and two do not so we may want to pick one whichever way we go. Or we don't have to be consistent since they all look slightly different anyway, with some all left aligned and some alternating, and different sizes of images! I don't care strongly just thought it might look nice.

Reviewer: Make a ticket for now and we can address it separately

Reviewer 2: Ticket made!

>       <p>Remotely control your rover base through the <a href="https://app.viam.com">Viam app.</a> Create an account and add a machine, <a href="/get-started/installation/">install viam-server</a>, and <a href="/components/base/">configure your base</a>. Then, go to the <b>Control</b> tab and access a remote control card for your base, with an interface for controlling speed, direction, and power.</p>

Reviewer 2: ```suggestion
      <p>You can remotely control your rover from anywhere through the <a href="https://app.viam.com">Viam app.</a> Create an account and add a machine, <a href="/get-started/installation/">install viam-server</a>, and <a href="/components/base/">configure your rover base</a>. Then, go to the <b>Control</b> tab and access a remote control card for your base, with an interface for controlling speed, direction, and power. You can also view live feeds from any cameras you configure.</p>
```
With the current imperative mood first sentence, it sounds like someone should start remotely controlling it already, and then make an account etc.

Reviewer: agreed. Was going to suggest something similar

    </th>
  </tr>
  <tr>
    <th>{{<imgproc src="/services/icons/base-rc.svg" class="fill alignleft" resize="200x" style="max-width: 200px" declaredimensions=true alt="Base remote control service icon.">}}
      <b>2. Program to move</b>

>       <p>Remotely control your rover base programmatically with a <a href="/sdks/">Viam SDK.</a> Make calls to the <a href="/mobility/motion/">motion service.</a>

Reviewer 2: ```suggestion
      <p>Remotely control your rover base programmatically with the <a href="/components/base/#api">base API</a>.
```
Take period out of link.

Motion service would be hard for most users to use effectively on a base without movement sensors and nav; let's take that out and focus on the path most users should take.

What do you think of linking directly to the base API so they can quickly see what they have available? Could see the argument for leaving as the SDKs page since that shows more different languages than the base API page currently does, but I think it'd help folks.

Reviewer: should the motion service then be mentioned with steps 3 and 4?

>       Or, <a href="/mobility/base-rc/">configure a base remote control service</a> to teleoperate your base with an <a href="/components/input-controller/">input controller.</a></p>

Reviewer 2: ```suggestion
      Or, <a href="/mobility/base-rc/">configure the base remote control service</a> to teleoperate your base with an <a href="/components/input-controller/">input controller.</a></p>
```
I think this sounds better/makes more sense since whereas there are lots of kinds of vision service, you'd only really be using one brc service

    </th>
  </tr>
  <tr>
    <th> {{<imgproc src="/services/icons/navigation.svg" class="fill alignright" resize="200x" style="max-width: 200px" declaredimensions=true alt="Navigation icon.">}}
      <b>3. Prepare your base to navigate</b>
      <p><a href="/components/movement-sensor/">Configure a movement sensor</a> as part of your machine to use navigation.

>       Then, <a href="/mobility/navigation/#configure-and-calibrate-the-frame-system-service-for-gps-navigation">configure and calibrate</a> the frame system for GPS navigation.</p>

Reviewer 2: Shouldn't "configure the navigation service" bit live up here?
I think you could make this section about all the configuration, and the next section about operation and monitoring.

    </th>
  </tr>
  <tr>
    <td>
    {{<imgproc src="/use-cases/navigation-card.png" class="fill alignleft" resize="200x" style="max-width: 300px" declaredimensions=true alt="Navigation map card">}}

>       <b>4. Navigate</b>

Reviewer 2: ```suggestion
      <b>4. Navigate autonomously</b>
```

>       <p><a href="/mobility/navigation/">Configure a navigation service</a> on your machine. Add waypoints to define a path. Manage navigation's start and stop, see where your machine is navigating, and configure waypoints and obstacles on the <b>Control</b> tab or with a <a href="/sdks/">Viam SDK.</a></p>

Reviewer 2: As above "*a* nav service" sounds odd to me--I think "the" makes more sense.

What does it mean to manage navigation's start and stop? Consider rewording.

If you can break out nav config into a previous section (per comment above), consider adding half a sentence more color here, such as *how* the control tab allows you to see where your machine is going, i.e. does it show a path in real time? is it a nice graphical map?
Remove duplication of adding waypoints.
Change sentence structure so that it's more clear whether you can do all these things from the control tab or SDK, or if you can only configure waypoints and obstacles with those.
How about something like
```suggestion
      <p>From the <b>Control</b> tab, use the map interface to add waypoints defining your machine's path. Then start your machine on its course and watch its position indicator move on the map. You can also control your robot's navigation programmatically with the <a href="/mobility/navigation/#api/">navigation API.</a></p>
```

Author: 1. 👍 
2. Start and stop navigating on the path, but its not necessarily stopping motion service so will rephrase. Maybe start and stop motion?

3. You can do both from both 

Good suggestion!

Once you have configured your machine, you can remotely control your machine on the app's **Control** tab, and set up autonomous navigation with the [navigation service](/mobility/navigation/).

<table>
  <tr>
    <th>{{<imgproc src="/use-cases/base-control.png" class="fill alignright" resize="200x" style="max-width: 200px" declaredimensions=true alt="Base control card">}}
      <b>1. Teleoperate</b>
      <p>You can remotely control your rover from anywhere through the <a href="https://app.viam.com">Viam app.</a> Create an account and add a machine, <a href="/get-started/installation/">install viam-server</a>, and <a href="/components/base/">configure your rover base</a>. Then, go to the <b>Control</b> tab and access a remote control card for your base, with an interface for controlling speed, direction, and power. You can also view live feeds from any cameras you configure.</p>
    </th>
  </tr>
  <tr>
    <th>{{<imgproc src="/services/icons/base-rc.svg" class="fill alignleft" resize="200x" style="max-width: 200px" declaredimensions=true alt="Base remote control service icon.">}}
      <b>2. Program to move</b>

>       <p>Remotely control your rover base programmatically with a <a href="/sdks/">Viam SDK.</a> Make calls to the <a href="/components/base/#api">base API</a> or <a href="/mobility/motion/">motion service</a>.

Author: @JessamyT @npentrel What do we think about motion service mention like this? It is hard to use but I agree w/ naomi it should still be on page

>       <p>Remotely control your rover base programmatically with a <a href="/sdks/">Viam SDK.</a> Make calls to the <a href="/components/base/#api">base API</a> or <a href="/mobility/motion/">motion service</a>.

Author: ```suggestion
      <p>Remotely control your rover base programmatically with a <a href="/sdks/">Viam SDK.</a> Make calls to the <a href="/components/base/#api">base API</a>.
```

      Or, <a href="/mobility/base-rc/">configure the base remote control service</a> to teleoperate your base with an <a href="/components/input-controller/">input controller.</a></p>
    </th>
  </tr>
  <tr>
    <th> {{<imgproc src="/services/icons/navigation.svg" class="fill alignright" resize="200x" style="max-width: 200px" declaredimensions=true alt="Navigation icon.">}}
      <b>3. Prepare your base to navigate</b>
      <p><a href="/components/movement-sensor/">Configure a movement sensor</a> as part of your machine to use navigation.
      Additionally, <a href="/mobility/navigation/#configure-and-calibrate-the-frame-system-service-for-gps-navigation">configure and calibrate</a> the frame system for GPS navigation.

>       Then, <a href="/mobility/navigation/">configure a navigation service</a> on your machine.</p>

Reviewer: You can currently have multiple navigation services

>       Then, <a href="/mobility/navigation/">configure a navigation service</a> on your machine.</p>

Reviewer 2: I was looking at https://viaminc.slack.com/archives/C0370C5QTCK/p1703103882874419?thread_ts=1703090068.507629&cid=C0370C5QTCK

>       Then, <a href="/mobility/navigation/">configure a navigation service</a> on your machine.</p>

Reviewer: ok !

    </th>
  </tr>
  <tr>
    <td>
    {{<imgproc src="/use-cases/navigation-card.png" class="fill alignleft" resize="200x" style="max-width: 300px" declaredimensions=true alt="Navigation map card">}}
      <b>4. Navigate autonomously</b>

>       <p> Define a path for your rover to navigate. Manage motion's start and stop, see where your machine is navigating, and configure waypoints and obstacles on the <b>Control</b> tab or with a <a href="/sdks/">Viam SDK</a>.</p>

Reviewer 2: It is still unclear with this sentence structure whether you can do all of the above with both, or can only configure waypoints and obstacles with both. I also still think it'd be good to give a little more description of the really nice tools we have for this process.

Also [nit] there's an extra space before "Define."

The following suggestion orders things such that configuring the waypoints and obstacles happens in the right sequence, before starting and stopping motion.

```suggestion
      <p>Define a path for your rover to navigate with waypoints and obstacles. Then, start and stop your machine's motion along the path and view your machine's current location. You can use the map interface on the <b>Control</b> tab or the <a href="/mobility/navigation/#api/">navigation API</a>.</p>
```

>       <p> Define a path for your rover to navigate. Manage motion's start and stop, see where your machine is navigating, and configure waypoints and obstacles on the <b>Control</b> tab or with a <a href="/sdks/">Viam SDK.</a></p>

Author: ```suggestion
      <p> Define a path for your rover to navigate. Manage motion's start and stop, see where your machine is navigating, and configure waypoints and obstacles on the <b>Control</b> tab or with a <a href="/sdks/">Viam SDK</a>.</p>
```

>       <p>Define a path for your rover to navigate with waypoints and obstacles. Then, start and stop your machine's motion along the path and view your machine's current location. You can use the map interface on the <b>Control</b> tab or the <a href="/mobility/navigation/#api/">navigation API</a>.</p>

Author: ```suggestion
      <p>Define a path for your rover to navigate with waypoints and obstacles. Then, start and stop your machine's motion along the path and view your machine's current location. You can use the map interface on the <b>Control</b> tab or the <a href="/mobility/navigation/#api">navigation API</a>.</p>
```
